### PR TITLE
[SPARK-26742][K8s][branch-2.4] Update k8s client version to 4.1.2

### PR DESCRIPTION
**Pull request by skonto**
## What changes were proposed in this pull request?

Updates client version and fixes some related issues.

## How was this patch tested?

Tested with the latest minikube version and k8s 1.13.
KubernetesSuite:
- Run SparkPi with no resources
- Run SparkPi with a very long application name.
- Use SparkLauncher.NO_RESOURCE
- Run SparkPi with a master URL without a scheme.
- Run SparkPi with an argument.
- Run SparkPi with custom labels, annotations, and environment variables.
- Run extraJVMOptions check on driver
- Run SparkRemoteFileTest using a remote data file
- Run SparkPi with env and mount secrets.
- Run PySpark on simple pi.py example
- Run PySpark with Python2 to test a pyfiles example
- Run PySpark with Python3 to test a pyfiles example
- Run PySpark with memory customization
- Run in client mode.
Run completed in 4 minutes, 20 seconds.
Total number of tests run: 14
Suites: completed 2, aborted 0
Tests: succeeded 14, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Spark Project Parent POM 2.4.2-SNAPSHOT ............ SUCCESS [  2.980 s]
[INFO] Spark Project Tags ................................. SUCCESS [  2.880 s]
[INFO] Spark Project Local DB ............................. SUCCESS [  1.954 s]
[INFO] Spark Project Networking ........................... SUCCESS [  3.369 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  1.791 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [  1.845 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  3.725 s]
[INFO] Spark Project Core ................................. SUCCESS [ 23.572 s]
[INFO] Spark Project Kubernetes Integration Tests 2.4.2-SNAPSHOT SUCCESS [04:25 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 05:08 min
[INFO] Finished at: 2019-03-06T18:03:55Z
[INFO] ------------------------------------------------------------------------

### CODE REVIEW

**Comment by skonto** [comment]
@shaneknapp I have this for 2.4.1 for testing as well. Tests fail due to the minikube issue.

**Comment by skonto** [comment]
jenkins test this please

**Comment by skonto** [comment]
I removed hadoop-2.6 deps but this still fails (although I try to align it with https://github.com/apache/spark/pull/23814)

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala [hunk 1]
```diff
@@ -26,8 +25,18 @@ import org.apache.spark.internal.Logging
...
+  private val KUBELET_PREFIX = "kubelet:"
+  private val APISERVER_PREFIX = "apiserver:"
+  private val KUBECTL_PREFIX = "kubectl:"
+  private val MINIKUBE_VM_PREFIX = "minikubeVM: "
+  private val MINIKUBE_PREFIX = "minikube: "
+  private val MINIKUBE_PATH = ".minikube"
+
+  def logVersion(): Unit = {
+    logInfo(executeMinikube("version").mkString("\n"))
+  }
```
Comment (skonto) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala]: we need this in order to support latest minikube versions.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala [hunk 2]
```diff
@@ -174,7 +174,7 @@ private[k8s] class LoggingPodStatusWatcherImpl(
         }.getOrElse(Seq(("Container state", "N/A")))
   }
 
-  private def formatTime(time: Time): String = {
-    if (time != null) time.getTime else "N/A"
+  private def formatTime(time: String): String = {
+    if (time != null ||  time != "") time else "N/A"
```
Comment (skonto) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala]: this part of the codebase has changed on master, but I am not porting back all the changes, just fixing this locally.

**Comment by Jeffwan** [comment]
@skonto Thanks! Good to know it's due to minikube issue. I was stucked on this for a while. Glad you figure it out

**Review by dongjoon-hyun — DISMISSED** [review]
Hi, @skonto . Please use the script to change dependency manifests.

```
$ dev/test-dependencies.sh --replace-manifest
```

Currently, the following is missing in this PR.
```
-logging-interceptor-3.8.1.jar
+logging-interceptor-3.12.0.jar
```

**Comment by skonto** [comment]
tests pass as reported here: https://issues.apache.org/jira/browse/SPARK-26742

**Comment by shaneknapp** [comment]
@gatorsmile -- am i missing something, or did this not get merged?

**Comment by shaneknapp** [comment]
ping @gatorsmile

**Comment by gatorsmile** [comment]
This is not merged. I accidentally closed it.

**Comment by shaneknapp** [comment]
k8s test failure is expected

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
we're going to wait to see if the regular non-k8s PRB job passes before merging.

**Comment by vanzin** [comment]
Merging to 2.4.

**Comment by yuhuali1989** [comment]
Hi，
I report an issue related with kubernetes-client 4.1.2 which introduce non daemon thread which block jvm exit.
Any Idea to fix it ? Thanks.

More details please check https://issues.apache.org/jira/browse/SPARK-27812

By the way, I believe update kubernetes-client is a necessary choice.
Without this version, I have to fix another issue with aws eks.

**Comment by Jeffwan** [comment]
@yuhuali1989 Could I know more details about issue with aws eks? 
I think maybe we can patch SDK client and then upgrade client in next 2.x release?

**Pull request closed without merging**